couchdb/server/views.py

```diff
@@ -0,0 +1,56 @@
+# -*- coding: utf-8 -*-
+#
+import copy
+import logging
+
+from couchdb import json
+from couchdb.server.exceptions import QueryServerException, Error
+
+__all__ = ('map_doc',)
+
+log = logging.getLogger(__name__)
+
+
+def map_doc(server, doc):
+    """Applies available map functions to document.
+
+    :command: map_doc
+
+    :param server: Query server instance.
+    :type server: :class:`~couchdb.server.BaseQueryServer`
+
+    :param doc: Document object.
+    :type doc: dict
+
+    :return: List of key-value results for each applied map function.
+
+    :raises:
+        - :exc:`~couchdb.server.exceptions.Error`
+          If any Python exception occurs due mapping.
+    """
+    docid = doc.get('_id')
+    log.debug('Apply map functions to document `%s`:\n%s', docid, doc)
+    orig_doc = copy.deepcopy(doc)
+    map_results = []
+    _append = map_results.append
+    try:
+        for idx, func in enumerate(server.state['functions']):
+            # TODO: https://issues.apache.org/jira/browse/COUCHDB-729
+            # Apply copy.deepcopy for `key` and `value` to fix this issue
+            _append([[key, value] for key, value in func(doc) or []])
+            if doc != orig_doc:
+                log.warning('Document `%s` had been changed by map function'
+                            ' `%s`, but was restored to original state',
+                            docid, func.__name__)
+                doc = copy.deepcopy(orig_doc)
+    except Exception as err:
+        msg = 'Exception raised for document `%s`:\n%s\n\n%s\n\n'
+        funsrc = server.state['functions_src'][idx]
+        log.exception(msg, docid, doc, funsrc)
+        if isinstance(err, QueryServerException):
+            raise
+        # TODO: https://issues.apache.org/jira/browse/COUCHDB-282
+        # Raise FatalError to fix this issue
+        raise Error(err.__class__.__name__, str(err))
```

Comment:
I will shelve the issue in TODO note ... until this PR get merged.

I think choosing `Error`/`FatalError` depend on user's senario. Everyone has different serverity level.

Someone may say, "My data is real-time data. They will expire in following few seconds. So, I do not care there are part of data do not work with `map` function.  Raising `Error` is fine and make qs still alive!"

Also, another senario could be, "I need to handle user registration data, all of them should not be dropped. Raising `FatalError` is worth for me. I can fix my code instantly."

Not sure where/how to implement this feature.
It should be implement as 'A parameter of view query url' ? , or 'A query server command line flag' ?, or (for example) 'magic var in mapfunction'? 
We need to discuss further.